sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.catalog;
+
+import org.apache.spark.annotation.Evolving;
+
+/**
+ * Represents a table which can be atomically truncated.
+ */
+@Evolving
+public interface TruncatableTable extends Table {
+  /**
+   * Truncate a table by removing all rows from the table atomically.
+   * If the table supports partitions, the method removes all existing partitions.
```

Comment:
Why is this required? A table may support empty partitions that exist independent of rows, like Hive tables. Is there a compelling reason to force this behavior?

Reply:
`SupportsPartitionManagement` has a `truncatePartition` method that will "Truncate a partition in the table by completely removing partition data."

That conflicts with the behavior of truncate here, which drops partitions. I think this requirement should be removed.

Reply:
> Is there a compelling reason to force this behavior?

Yes, I wanted to highlight that it can remove not only all rows but also partitions to align to Spark's v1 (and Hive) `TRUNCATE TABLE` but the command doesn't drop partitions in fact:
```sql
spark-sql> CREATE TABLE tbl (col0 INT) PARTITIONED BY (part INT);
spark-sql> INSERT INTO tbl PARTITION (part=0) SELECT 0;
spark-sql> ALTER TABLE tbl ADD PARTITION (part=1);
spark-sql> SHOW PARTITIONS tbl;
part=0
part=1
spark-sql> SELECT * FROM tbl;
0	0
spark-sql> TRUNCATE TABLE tbl;
spark-sql> SHOW PARTITIONS tbl;
part=0
part=1
spark-sql> SELECT * FROM tbl;
spark-sql>
```
Let me remove the sentence.

Reply:
but in this PR, I drop empty partitions as well. The question is should v2 `TRUNCATE TABLE` be aligned to v1 implementation, and preserve empty partitions? I guess the answer is yes.

Reply:
Surprising but we don't have any tests that checks partition existence after entire table truncation. I opened this PR https://github.com/apache/spark/pull/31544 with such checks for tables from v1 In-Memory and Hive external catalogs.